pkg/resource/bucket/hook.go

```diff
@@ -276,7 +277,42 @@ func customPreCompare(
 		if b.ko.Spec.ACL != nil {
 			b.ko.Spec.ACL = matchPossibleCannedACL(*a.ko.Spec.ACL, *b.ko.Spec.ACL)
 		}
+	} else {
+		// If we are sure the grants weren't set from the header strings
+		if a.ko.Spec.GrantFullControl == nil &&
+			a.ko.Spec.GrantRead == nil &&
+			a.ko.Spec.GrantReadACP == nil &&
+			a.ko.Spec.GrantWrite == nil &&
+			a.ko.Spec.GrantWriteACP == nil {
+			b.ko.Spec.GrantFullControl = nil
+			b.ko.Spec.GrantRead = nil
+			b.ko.Spec.GrantReadACP = nil
+			b.ko.Spec.GrantWrite = nil
+			b.ko.Spec.GrantWriteACP = nil
+		}
+
+		emptyGrant := ""
+		if a.ko.Spec.GrantFullControl == nil && b.ko.Spec.GrantFullControl != nil {
+			a.ko.Spec.GrantFullControl = &emptyGrant
+			// TODO(RedbackThomson): Remove the following line. GrantFullControl
+			// has a server-side default of id="<owner ID>". This field needs to
+			// be marked as such before we can diff it.
+			b.ko.Spec.GrantFullControl = &emptyGrant
+		}
+		if a.ko.Spec.GrantRead == nil && b.ko.Spec.GrantRead != nil {
+			a.ko.Spec.GrantRead = &emptyGrant
+		}
+		if a.ko.Spec.GrantReadACP == nil && b.ko.Spec.GrantReadACP != nil {
+			a.ko.Spec.GrantReadACP = &emptyGrant
+		}
+		if a.ko.Spec.GrantWrite == nil && b.ko.Spec.GrantWrite != nil {
+			a.ko.Spec.GrantWrite = &emptyGrant
+		}
+		if a.ko.Spec.GrantWriteACP == nil && b.ko.Spec.GrantWriteACP != nil {
+			a.ko.Spec.GrantWriteACP = &emptyGrant
+		}
```

Comment:
we might consider placing the above into a little utility function, ala `matchPossibleCannedACL`. But you can handle that in a future PR if you want...

Reply:
I'm still not certain my logic here is flawless. Too many code paths to mentally track for me to be confident in it. I probably need to add some integration tests that switch between the types.